### PR TITLE
Fall back to read/write if sendfile fails with EINVAL or ENOSYS

### PR DESCRIPTION
**Pull request by hmoffatt**
Fix for microsoft/vcpkg#24794

### CODE REVIEW

**Comment by BillyONeal** [comment]
Thanks so much for the fix!

File: src/vcpkg/base/files.cpp [hunk 1]
```diff
@@ -2887,6 +2887,16 @@ namespace vcpkg
...
+#if defined(__APPLE__)
+            if (fcopyfile(source_fd.get(), destination_fd.get(), 0, COPYFILE_ALL) == -1)
+            {
+                ec.assign(errno, std::generic_category());
+                return false;
+            }
+
+            // fcopyfile copies the mode so no need to fchmod here
+            return true;
+#else
```
Comment (dg0yt) [inline src/vcpkg/base/files.cpp:2899]: `#else // ^^^ defined(__APPLE__) / !(defined(__APPLE__) vvv`

File: src/vcpkg/base/files.cpp [hunk 2]
```diff
@@ -2901,24 +2911,28 @@ namespace vcpkg
...
+            if (!ec)
             {
-                ec.assign(errno, std::generic_category());
-                return false;
+                destination_fd.fchmod(source_stat.st_mode, ec);
+                return true;
             }
+            // Else fall back to read/write
+            ec.clear();
+#endif // ^^^ !(defined(__APPLE__) || defined(__linux__))
```
Comment (dg0yt) [inline src/vcpkg/base/files.cpp:2934]: `#endif  // ^^^ (defined(__linux__)`

**Review by dg0yt — COMMENTED** [review]
Do the comments need modifications?

File: src/vcpkg/base/files.cpp (hunk 2 above)
Comment (hmoffatt) [reply src/vcpkg/base/files.cpp:2934]: I have just submitted #544 to fix those.

**Pull request merged**